### ISSUE TITLE
Map Matching: buggy behaviour when making a uturn on virtual edges

### ISSUE THREAD

**Issue by YousefED** [graphhopper/graphhopper#2613]
**Describe the bug**
I've encountered a bug in Map Matching where 1 point that goes back can cause large, unintended consequences.

**To Reproduce**

I'm sending a simple route to the map matching API:

<img width="639" alt="image" src="https://user-images.githubusercontent.com/368857/175829352-22202381-92a7-4e51-be0d-04528d965a4c.png">

It consists of a few points on a single road. There is one point where the user went back (the point in the middle) before continuing the road regularly (i.e.: two u-turns). See the attached geojson or gpx files for the exact data

1. Send this to the Map Matching API

```
curl -XPOST -H "Content-Type: application/gpx+xml" "https://graphhopper.com/api/1/match?profile=foot&key=API_KEY&gps_accuracy=5&points_encoded=false" --data @input.gpx
```
(gps accuracy is set to 5 meters as the path is drawn closely to the real road)

2. inspect results:

<img width="668" alt="image" src="https://user-images.githubusercontent.com/368857/175829464-be067a76-da86-496d-b51f-5a1194228548.png">

The black part is the input, the red line is the output from the API. (note: the red line goes out-and back to the end node, twice over the same path, which is not possible to see from the screenshot). The resulting route is way longer than the input.

**Expected behavior**
I would expect one of these two scenarios to happen:

a. The Matched path follows the user exactly, with an "inner u-turn"
b. The u-turn is ignored and the Matched path is just a subpart of the road without the u-turn

Scenario A would make most sense, with gps_accuracy=5 I'd expect to see the u-turn actually show up in the results

**Possibly related**
Perhaps the behaviour we're seeing here is an unintended consequence of https://github.com/graphhopper/map-matching/pull/83, but I might be wrong (relatively old PR).

**Comment by easbar** [comment]
Do you get the same result when you add `lm.disable=true` to your request?

**Comment by YousefED** [comment]
> Do you get the same result when you add `lm.disable=true` to your request?

Yes, both appear to give the exact same results

**Comment by YousefED** [comment]
We noticed that changing:

`createRouter().calcPath(from.getSnap().getClosestNode(), to.getSnap().getClosestNode(), from.isOnDirectedEdge() ? from.getOutgoingVirtualEdge().getEdge() : EdgeIterator.ANY_EDGE, to.isOnDirectedEdge() ? to.getIncomingVirtualEdge().getEdge() : EdgeIterator.ANY_EDGE);`

to 

`createRouter().calcPath(from.getSnap().getClosestNode(), to.getSnap().getClosestNode(), EdgeIterator.ANY_EDGE, EdgeIterator.ANY_EDGE);`

in the MapMatching code (i,e: remove the directededge options) fixes the above issue. However, as it's quite a complicated algorithm overall it's not trivial to oversee all consequences of this change. I suppose edge directions are in there for a good reason. Do you think this approach is viable or can you point me to what in your opinion might be a better approach to fix this issue? (would be happy to test and contribute a fix)

**Comment by michaz** [comment]
> Perhaps the behaviour we're seeing here is an unintended consequence of https://github.com/graphhopper/map-matching/pull/83, but I might be wrong (relatively old PR).

Correct, it's an unintended (but anticipated) consequence of that PR, which was fixing a worse problem though. Back then, many things were not possible or very complicated to do with GraphHopper, most notably representing the side of the road you **are currently on** in a routing state so that it works with all algorithms and in all edge cases.

This is AFAIK all resolved in the meantime, but things remained at "inner-link U-turns are forbidden now" because that was what was both desirable and possible at that time and we didn't revisit it since then.

I'll have a look in the coming days.

**Comment by dteirney** [comment]
@michaz we think we've run into this same issue with inner-link U-turns that u-turn at the next intersection as opposed to the GPS location where the u-turn around happened.

> Of course, people do actually make "inner-link U turns" all the time so it must be possible for a correct map-matching to pick them up.
> 
> I'll have a look in the coming days.

Did you ever manage to make time to have a look all that time ago?

**Comment by michaz** [comment]
Well days turned into years, and a certain technological change was required for me to pick up speed on this, but it's fixed now, since 25494d28.
